tensorhive/controllers/reservation.py

```diff
@@ -96,66 +98,64 @@ def create(reservation: Dict[str, Any]) -> Tuple[Content, HttpStatusCode]:
     except AssertionError as e:
         content = {'msg': RESERVATION['create']['failure']['invalid'].format(reason=e)}
         status = 422
-    except Exception:
-        content = {'msg': GENERAL['internal_error']}
+    except Exception as e:
+        print(e)
+        content = {'msg': GENERAL['internal_error'] + str(e)}
         status = 500
     finally:
         return content, status
 
 
-def to_db_column() -> Dict[str, str]:
-    return {
-        'title': 'title',
-        'description': 'description',
-        'resourceId': 'protected_resource_id',
-        'start': 'starts_at',
-        'end': 'ends_at',
-    }
-
-
 @jwt_required
 def update(id: ReservationId, newValues: Dict[str, Any]) -> Tuple[Content, HttpStatusCode]:
     new_values = newValues
-    allowed_fields = {'title', 'description', 'resourceId', 'start', 'end'}
+    allowed_fields = {'title', 'description', 'resourceId', 'end'}
     try:
-        assert set(new_values.keys()).issubset(allowed_fields), 'invalid field is present'
         reservation = Reservation.get(id)
 
+        if reservation.end < datetime.datetime.utcnow() and not is_admin():
+            raise ForbiddenReservationException('reservation already finished')
+
+        if reservation.start > datetime.datetime.utcnow() or is_admin():
+            allowed_fields.add('start')
+
+        if not set(new_values.keys()).issubset(allowed_fields):
+            raise ForbiddenReservationException('invalid field is present')
+
         for field_name, new_value in new_values.items():
-            field_name = to_db_column().get(field_name)
+            field_name = snakecase(field_name)
             assert (field_name is not None) and hasattr(reservation, field_name), \
                 'reservation has no {} field'.format(field_name)
             setattr(reservation, field_name, new_value)
 
-        assert __is_admin_or_reservation_owner(reservation), GENERAL['unprivileged']
-
         user = User.get(get_jwt_identity())
-        if ReservationVerifier.is_reservation_allowed(user, reservation):
-            reservation.save()
-            content, status = {'msg': RESERVATION['update']['success'], 'reservation': reservation.as_dict}, 201
-        else:
-            content, status = {'msg': RESERVATION['update']['failure']['forbidden']}, 403
+        if not ReservationVerifier.is_reservation_allowed(user, reservation) or not \
+                (is_admin() or __is_reservation_owner(reservation)):
```

Comment:
```suggestion
        if not (is_admin() or __is_reservation_owner(reservation))  or not \
                ReservationVerifier.is_reservation_allowed(user, reservation):
```
I missed it last time. It is also worth changing to make it faster.